xen/arch/arm/coproc/schedule.c

```diff
@@ -0,0 +1,317 @@
+	}
+	vcoproc->state = VCOPROC_UNKNOWN;
+	spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+
+	return 0;
+}
+
+void vcoproc_sheduler_vcoproc_wake(struct vcoproc_scheduler *sched, struct vcoproc_instance *vcoproc)
+{
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+	unsigned long flags;
+
+	ASSERT(!vcoproc_scheduler_vcoproc_is_destroyed(sched, vcoproc));
+
+	/* TODO What to do if we came with state TERMINATING? */
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+	if ( vcoproc->state != VCOPROC_SLEEPING )
+	{
+		spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+		return;
+	}
+
+	VCOPROC_SCHED_OP(sched, wake, vcoproc);
+	vcoproc->state = VCOPROC_WAITING;
+	spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+
+	vcoproc_schedule(sched);
+}
+
+void vcoproc_sheduler_vcoproc_sleep(struct vcoproc_scheduler *sched, struct vcoproc_instance *vcoproc)
+{
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+	unsigned long flags;
+	int reschedule = 0;
+
+	ASSERT(!vcoproc_scheduler_vcoproc_is_destroyed(sched, vcoproc));
+
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+	if ( vcoproc->state != VCOPROC_WAITING && vcoproc->state != VCOPROC_RUNNING )
+	{
+		spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+		return;
+	}
+
+	VCOPROC_SCHED_OP(sched, sleep, vcoproc);
+	if ( vcoproc->state == VCOPROC_WAITING )
+		vcoproc->state = VCOPROC_SLEEPING;
+	else
+	{
+		vcoproc->state = VCOPROC_TERMINATING;
+		reschedule = 1;
+	}
+	spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+
+	if ( reschedule )
+		vcoproc_schedule(sched);
+}
+
+void vcoproc_sheduler_vcoproc_yield(struct vcoproc_scheduler *sched, struct vcoproc_instance *vcoproc)
+{
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+	unsigned long flags;
+
+	ASSERT(!vcoproc_scheduler_vcoproc_is_destroyed(sched, vcoproc));
+
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+	if ( vcoproc->state != VCOPROC_RUNNING )
+	{
+		spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+		return;
+	}
+	VCOPROC_SCHED_OP(sched, yield, vcoproc);
+	spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+
+	vcoproc_schedule(sched);
+}
+
+/* TODO Taking lock for the whole func is might be an overhead */
+void vcoproc_schedule(struct vcoproc_scheduler *sched)
+{
+	struct vcoproc_instance *curr, *next;
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+	struct vcoproc_task_slice next_slice;
+	s_time_t now;
+	unsigned long flags;
+
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+
+	curr = vcoproc_scheduler_get_current(sched);
+
+	now = NOW();
+	stop_timer(&sched_data->s_timer);
+
+	next_slice = sched->do_schedule(sched, now);
+	next = next_slice.task;
+
+	if ( unlikely(!curr && !next) )
+		goto out;
+
+	if ( vcoproc_context_switch(curr, next) )
+	{
+		if ( coproc_wait_time >= 0 )
+			set_timer(&sched_data->s_timer, now + coproc_wait_time);
+
+		VCOPROC_SCHED_OP(sched, schedule_completed, next, 0);
```

Comment:
Yes, we failed to switch to "next". So, we have to put "next" back to the runqueue head.
For doing this we have to pass "0" thru last param.